internal/impl/sftp/input.go

```diff
@@ -173,180 +174,209 @@ func newSFTPReaderFromParsed(conf *service.ParsedConfig, mgr *service.Resources)
-	s.scannerMut.Lock()
-	scanner := s.scanner
-	s.scanner = nil
-	client := s.client
-	s.client = nil
-	s.paths = nil
-	s.scannerMut.Unlock()
-
-	if scanner != nil {
-		if err := scanner.Close(ctx); err != nil {
-			s.log.With("error", err).Warn("Failed to close consumed file")
+func (s *sftpReader) initScanner(ctx context.Context, client *sftp.Client) error {
+	if s.scanner != nil {
+		return nil
+	}
+
+	var file *sftp.File
+	var path string
+	for {
+		var ok bool
+		var err error
+		path, ok, err = s.pathProvider.Next(ctx, client)
+		if err != nil {
+			return fmt.Errorf("finding next file path: %w", err)
+		}
+		if !ok {
+			return service.ErrEndOfInput
+		}
+
+		file, err = client.Open(path)
+		if err != nil {
+			s.log.With("path", path, "err", err.Error()).Warn("Unable to open previously identified file")
+			if os.IsNotExist(err) {
+				// If we failed to open the file because it no longer exists then we
+				// can "ack" the path as we're done with it. Otherwise we "nack" it
+				// with the error as we'll want to reprocess it again later.
+				err = nil
+			}
+			if ackErr := s.pathProvider.Ack(ctx, path, err); ackErr != nil {
+				s.log.With("error", ackErr).Warnf("Failed to acknowledge path: %s", path)
+			}
+			continue
 		}
+
+		details := service.NewScannerSourceDetails()
+		details.SetName(path)
+		scanner, err := s.scannerCtor.Create(file, s.newCodecAckFn(path), details)
+		if err != nil {
+			_ = file.Close()
+			return fmt.Errorf("creating scanner: %w", err)
+		}
+
+		s.scanner = scanner
+		s.currentPath = path
+		return nil
 	}
-	if client != nil {
-		if err := client.Close(); err != nil {
-			s.log.With("error", err).Error("Failed to close client")
+}
+
+func (s *sftpReader) newCodecAckFn(path string) service.AckFunc {
+	return func(ctx context.Context, aErr error) error {
+		s.stateLock.Lock()
+		defer s.stateLock.Unlock()
+
+		if err := s.pathProvider.Ack(ctx, path, aErr); err != nil {
+			s.log.With("error", err).Warnf("Failed to acknowledge path: %s", path)
+		}
+		if aErr != nil {
+			return nil
+		}
+
+		if s.deleteOnFinish {
+			client, cleanup, err := s.initClient()
+			if err != nil {
+				return err
+			}
+			defer cleanup()
+
+			if err := client.Remove(path); err != nil {
+				return fmt.Errorf("remove %v: %w", path, err)
+			}
 		}
+
+		return nil
 	}
+}
+
+func (s *sftpReader) Close(ctx context.Context) error {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	s.closeScanner(ctx)
+	s.closeClient()
+	s.closed = true
 	return nil
 }
 
-//------------------------------------------------------------------------------
+func (s *sftpReader) closeScanner(ctx context.Context) {
+	if s.scanner != nil {
+		if err := s.scanner.Close(ctx); err != nil {
+			s.log.With("error", err).Error("Failed to close scanner")
+		}
+		s.scanner = nil
+		s.currentPath = ""
+	}
+}
 
-var errEndOfPaths = errors.New("end of paths")
+func (s *sftpReader) closeClient() {
+	if s.client == nil {
```

Comment:
Not sure if it's an issue, but when `s.closed == true`, can we get multiple concurrent calls to this `closeClient()` method via the `initClient()` `cleanup` which is called in the ackFn callback. If yes, I'm thinking that one goroutine could close the client while the other one is just preparing to remove the file.